File: getData.py

#import requests;
import json;
import asyncio;
import aiohttp;
import traceback;
import re;
from pathlib import Path;
from collections import Counter;


#FUNCTIONS
def readKeywords(path):
    techList = []   
    """Read keywords from a text file where each value is on a seperate line"""
    try:
        with open(path) as input_file:
            #stripping newline characters
            techList = [line.rstrip('\n') for line in input_file]
            techList = [line.upper() for line in techList]


    #if an error occurs, print the error.
    except Exception as e: print("Problem with technologies.txt\n--------------------------\n" + str(e))
    if len(techList) < 1:
        raise Exception("No keywords found")
    return techList

async def parseData(postings):
    """Fetch json, parse the data and prepare it to be turned into json"""

    keywordMentions = []

    for i in postings:
        s = i['descr'].upper()
        heading = i["heading"]
        company = i["company_name"]
        #Remove zipcodes and '-' signs using regex
        location = re.sub(r"\d+|-+", '', str(i["municipality_name"])).strip()
        if(location == ""):
            location = "none";
        link = "https://duunitori.fi/tyopaikat/tyo/" + i["slug"]

        #List of keywords found in post
        regexFound = re.findall(r"\bC\+\+|\bC\#|\b" + r'\b|\b'.join(KWList) + r"\b", s, re.IGNORECASE);
        keywordsFound = list(set(regexFound))

        #keywordsFound = list({x for x in KWList if bool(re.search(r"\b" + x + r"\b", s, re.IGNORECASE))})

        for i in range(len(keywordsFound)):
            #strip whitespace and remove , and ) characters from the matches
            keywordsFound[i] = keywordsFound[i].strip()
            keywordsFound[i] = keywordsFound[i].replace(",", "").replace(")", "").replace("-", "").replace("(", "")

        job = {"heading": heading, "link": link, "technologies": list(set(keywordsFound)), "company": company, "location": location}

        for keyword in keywordsFound:
            if (keyword in technologies):
                technologies[keyword]['jobs'].append(job)
            else:
                technologies[keyword] = dict()
                technologies[keyword]['name'] = keyword
                technologies[keyword]['jobs'] = []
                technologies[keyword]['jobs'].append(job)


        #create company and location dictionaries
        if len(keywordsFound) > 0:
            if company in companies:
                companies[company]["technologies"] += keywordsFound
                companies[company]["jobs"].append(job)

            else:
                companies[company] = dict()
                companies[company]["name"] = company
                companies[company]["technologies"] = keywordsFound
                companies[company]["jobs"] = []
                companies[company]["jobs"].append(job)

            if location in locations:
                locations[location]["technologies"] += keywordsFound
                locations[location]["jobs"].append(job)
            else:
                locations[location] = dict()
                locations[location]["name"] = location
                locations[location]["technologies"] = keywordsFound
                locations[location]["jobs"] = []
                locations[location]["jobs"].append(job)

            keywordMentions += keywordsFound
    return keywordMentions

#Fetch the postings site json
async def fetch(url):
    """Execute an http call async
    Args:
        url: URL to call
    Return:
        responses: A dict like object containing http response
    """
    async with aiohttp.ClientSession(headers={"accept": "application/json", "User-Agent": "Curl/7.64.1"}) as session:
        async with session.get(url) as response:
            resp = await response.json()
            print(resp)
            return resp

#fetch all postings
async def handleData():
    tasks = []
    allMentionsFound = []
    #call api to get all postings
    url = "https://duunitori.fi/api/v1/jobentries?search=koodari&search_also_descr=1&format=json"
    results = await fetch(url) 
    print("results gotten from: " + url)
    count = results["count"]
    posts = results["results"]
    tasks.append(parseData(results['results']))
    nextPage = results['next']

    while nextPage != None:
        url = nextPage
        results = await fetch(url)
        print("results gotten from: " + url)
        posts += results["results"]
        tasks.append(parseData(results['results']))
        nextPage = results['next']


    jsonDataToFile({"count": count, "posts": posts}, Path('./json/posts.json'))

    mentions = (await asyncio.gather(*tasks, return_exceptions=False))
    for m in mentions:
        allMentionsFound += m


    return allMentionsFound, count

def formatAndCreateJson(count, kwMentions):
    """Format and create a usable json file from data"""
    global companies
    global locations
    techCount = dict()
    techCount = Counter(kwMentions)
    for key in techCount:
        technologies[key]['jobs_count'] = techCount[key]

    for key in companies:
        companies[key]["technologies"] = Counter(companies[key]["technologies"])
        companies[key]["jobs_count"] = len(companies[key]["jobs"])


    for key in locations:
        locations[key]["technologies"] = Counter(locations[key]["technologies"])
        locations[key]["jobs_count"] = len(locations[key]["jobs"])
    
    data["posts_count"] = count
    data["technologies"] = technologies
    data["companies"] = companies
    data["locations"] = locations

    jsonDataToFile(data, Path('./json/data.json'))




def jsonDataToFile(d, path):
    """Creates/overwrites data to a file in json format"""
    try:
        jsonData = json.dumps(d)

        f = open(path, "w+")
        f.write(jsonData)
    except Exception as e: print(f"An exception occurred while saving json data to file {path}:\n {e}")


#VARIABLES
KWList = []
technologies = dict()
companies = dict()
locations = dict()
data = dict()

try:
    KWList = readKeywords(Path('./keywords/technologies.txt'))
    allMentions, count = asyncio.run(handleData())
    formatAndCreateJson(count, allMentions)
except Exception:
    print(traceback.format_exc())

File: getData.go
package main

import (
    "fmt"
    "os"
    "io/ioutil"
    "strings"
    "regexp"
    "net/http"
    "encoding/json"
    "github.com/lassi-koykka/fin-dev-api/set"
)


const BASE_URL = "https://duunitori.fi/api/v1/jobentries?search=koodari&search_also_descr=1&format=json"

type Posting struct {
    Slug string
    Heading string
    Date_posted string
    Municipality_name *string
    Export_image_url *string
    Company_name string
    Descr string
}

type ApiData struct {
    Count int
    Next *string
    Previous *string
    Results []Posting
}

func check (err error) {
    if err != nil {
        fmt.Println(err)
        panic(err)
    }
}
func readKeywords() []string {
    data, err := os.ReadFile("keywords/technologies.txt")
    check(err)
    content := string(data)
    keywords := strings.Split(content, "\n")
    result := []string{}
    for _, kw := range keywords {
        result = append(result, strings.TrimSpace(kw))
    }
    return keywords
}

func readPostsFromFile() ApiData {
    data, err := os.ReadFile("posts.json")
    check(err)
    return ParseApiData(data)
}

func GetApiData(url string) []byte {
    res, err := http.Get(url)
    check(err)
    defer res.Body.Close()
    bodyData, parseBodyErr := ioutil.ReadAll(res.Body)
    check(parseBodyErr)
    return bodyData
}

func ParseApiData(bodyData []byte) ApiData {
    var data ApiData
    err := json.Unmarshal(bodyData, &data)
    check(err)
    return data
}

func getAllPosts(saveToFile bool) ApiData {
    bodyData := GetApiData(BASE_URL)

    var data ApiData
    data = ParseApiData(bodyData)

    next := data.Next


    i := 1
    for next != nil {
        println("GET " + *next)
        fmt.Printf("Try %d \n", i)
        var newData ApiData
        newBodyData := GetApiData(*next)
        newData = ParseApiData(newBodyData)
        data.Results = append(data.Results, newData.Results...)
        next = newData.Next
        i++
    }

    if saveToFile {
        res, err := json.Marshal(data)
        check(err)
        writeErr := os.WriteFile("posts.json", res, 0666)
        check(writeErr)
    }

    return data
}

// Go through each posting and
// 1. regex all match all words in keywords list
func handleData(postings []Posting, keywords []string) {
    keywordsFound := set.NewStrSet()
    technologiesByPosting := make(map[string][]string)
    //TODO FINISH COUNTING
    // technologyCountsOverall := make(map[string]int)
    // technologyCountsByCity := make(map[string]map[string]int)
    for _, p := range postings {
        descr := strings.ToLower(p.Descr)
        for _, keyword := range keywords {
            kw := strings.ToLower(keyword)
            resultKw := strings.ReplaceAll(keyword, ".js", "")
            found := false
            if len(kw) == 1 {
                matchString := `\b(` + kw + `\.|` + kw + `\,)\b`
                result, err := regexp.MatchString(matchString, descr)
                check(err)
                found = result
            } else if strings.Contains(kw, ".js") {
                matchString := `\b(` + kw + "|" + strings.ReplaceAll(kw, ".js", "") + `)\b`
                result, err := regexp.MatchString(matchString, descr)
                check(err)
                found = result
            } else if strings.ContainsAny(kw, "#+.- ") {
                if strings.Contains(descr, kw) { found = true }
            } else {
                matchString := `\b(` + kw + `)\b`
                result, err := regexp.MatchString(matchString, descr)
                check(err)
                found = result 
            }

            if found { 
                keywordsFound.Add(resultKw) 
            }

        }

        // city := strings.ToLower(*p.Municipality_name)
        fmt.Println(keywordsFound.ToSlice())
        technologiesByPosting[p.Slug] = keywordsFound.ToSlice()
    }
}

func main() {
    println("Running main")
    keywords := readKeywords()

    // apiData := getAllPosts(true)
    apiData := readPostsFromFile()
    handleData(apiData.Results, keywords)
    fmt.Println(len(apiData.Results))
}
